### DISCUSSION TITLE
How to start as a background service?

### DISCUSSION THREAD

**Discussion by dentech82** [DennisOSRM/hms-mqtt-publisher#95] [General]
Hello,

the script is running as cli tool and not as a background service. How can I run the script in background?

OS: Raspbian in 64 bit

Greets

**Comment by l0tzi** [comment]
Hi,
I think the best option is to create a systemd service for the tool. You can check i.e. [this post](https://medium.com/@benmorel/creating-a-linux-service-with-systemd-611b5c8b91d6) for a reference

**Reply by dentech82** [reply]
Thanks, I got it.

**Reply by ghost** [reply]
@dentech82 How did you manage to run it as a background service?

**Reply by dentech82** [reply]
Hey,

i've added a script to /etc/systemd/system/ , like hms-publish.service :

```
[Unit]
Description=hms-mqtt publisher
After=network.target
StartLimitIntervalSec=0
[Service]
Type=simple
Restart=always
RestartSec=1
User=data
ExecStart=/home/data/hms-mqtt-publisher/target/release/hms-mqtt-publish

[Install]
WantedBy=multi-user.target
```

Start the service with: "systemctl start hms-mqtt-publish"
and for automatically start on boot "systemctl enable hms-mqtt-publish"

that's it.

**Reply by Diemo79** [reply]
Good evening, I hope you can help me out. 
I created a service file "hms-publish.service" in /etc/systemd/system/ I can run the start command without getting an error message, but unfortunately no service is active. :-( What is my mistake?

hms-publish.service:

[Unit]
Description=hms-mqtt publisher
After=network.target
StartLimitIntervalSec=0
[Service]
Type=simple
Restart=always
RestartSec=1
User=user79
ExecStart=/home/user79/hms-mqtt-publisher/target/debug/hms-mqtt-publish

[Install]
WantedBy=multi-user.target

**Reply by ghost** [reply]
@Diemo79 I guess your ExecStart path is wrong. You need it to change to: `/home/user79/hms-mqtt-publisher/target/release/hms-mqtt-publish`

**Reply by Diemo79** [reply]
![image](https://github.com/user-attachments/assets/5b06988f-9065-49c9-bab9-a5accdf2df43)

Thanks for your answer. Under the folder "Target" is just one folder --> "debug"

**Reply by Diemo79** [reply]
![image](https://github.com/user-attachments/assets/308ee2c1-dcc2-459a-8cba-f37aa42b6e20)

**Reply by Diemo79** [reply]
user79@raspi:/etc/systemd/system $ cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"

**Reply by Diemo79** [reply]
user79@raspi:/etc/systemd/system $ systemctl start hms-publish.service
==== AUTHENTICATING FOR org.freedesktop.systemd1.manage-units ===
Authentication is required to start 'hms-publish.service'.
Authenticating as: ,,, (user79)
Password:
==== AUTHENTICATION COMPLETE ===
user79@raspi:/etc/systemd/system $ systemctl status hms-publish.service
● hms-publish.service - hms-mqtt publisher
     Loaded: loaded (/etc/systemd/system/hms-publish.service; disabled; vendor preset: enabled)
     Active: activating (auto-restart) (Result: exit-code) since Sun 2024-07-28 13:21:16 CEST; 560ms ago
    Process: 14534 ExecStart=/home/user79/hms-mqtt-publisher/target/release/hms-mqtt-publish (code=exited, status=203/EXEC)
   Main PID: 14534 (code=exited, status=203/EXEC)
        CPU: 4ms

**Reply by ghost** [reply]
@Diemo79 Run `cargo build --release`. This should create the release folder

**Reply by Diemo79** [reply]
user79@raspi:/etc/systemd/system $ systemctl status hms-publish.service
● hms-publish.service - hms-mqtt publisher
     Loaded: loaded (/etc/systemd/system/hms-publish.service; disabled; vendor preset: enabled)
     Active: activating (auto-restart) (Result: exit-code) since Sun 2024-07-28 13:30:03 CEST; 248ms ago
    Process: 18827 ExecStart=/home/user79/hms-mqtt-publisher/target/debug/hms-mqtt-publish (code=exited, status=101)
   Main PID: 18827 (code=exited, status=101)
        CPU: 14ms

**Reply by Diemo79** [reply]
Hello, thank you for your patient support.
Unfortunately I'm still having problems starting the background service. What I did: 
- Configured "config.toml"
1. Execute the following command in the "hms-mqtt-publisher" folder: "cargo r" --> Result: OK (data are fetched and forwarded)
2. Execute the following command in the "hms-mqtt-publisher" folder: "cargo r build --release" --> Result: OK (data are fetched and forwarded, release folder including files is created)
3. Script file created in /etc/systemd/system with the following content.

[Unit]
Description=hms-mqtt publisher
After=network.target
StartLimitIntervalSec=0
[Service]
Type=simple
Restart=always
RestartSec=1
User=user79
ExecStart=/home/user79/hms-mqtt-publisher/target/release/hms-mqtt-publish

[Install]
WantedBy=multi-user.target


 
Unfortunately without success after the service started.  With ...status... I get the following message. 

![image](https://github.com/user-attachments/assets/5a6d6798-e1b6-4f4c-b8e6-f4475c54419c)


Any idea?

**Reply by Diemo79** [reply]
Hi, I found the error. The WorkingDirectory was missing in the "*.Service" file. Below is the now working script.

![image](https://github.com/user-attachments/assets/24d5d8b2-a4fb-4d1f-93b6-6987086a6ac0)
